x-pack/plugins/security_solution/server/lib/siem_migrations/rules/api/util/prebuilt_rules.ts

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { SavedObjectsClientContract } from '@kbn/core/server';
+import type { RulesClient } from '@kbn/alerting-plugin/server';
+import type { RuleResponse } from '../../../../../../common/api/detection_engine';
+import { createPrebuiltRuleObjectsClient } from '../../../../detection_engine/prebuilt_rules/logic/rule_objects/prebuilt_rule_objects_client';
+import { fetchRuleVersionsTriad } from '../../../../detection_engine/prebuilt_rules/logic/rule_versions/fetch_rule_versions_triad';
+import { createPrebuiltRuleAssetsClient } from '../../../../detection_engine/prebuilt_rules/logic/rule_assets/prebuilt_rule_assets_client';
+import { convertPrebuiltRuleAssetToRuleResponse } from '../../../../detection_engine/rule_management/logic/detection_rules_client/converters/convert_prebuilt_rule_asset_to_rule_response';
+
+interface PrebuiltRulesResults {
+  /**
+   * The latest available version
+   */
+  target: RuleResponse;
+
+  /**
+   * The currently installed version
+   */
+  current?: RuleResponse;
+}
+
+export const getPrebuiltRules = async (
+  rulesClient: RulesClient,
+  savedObjectsClient: SavedObjectsClientContract,
+  rulesIds?: string[]
+): Promise<PrebuiltRulesResults[]> => {
+  const ruleAssetsClient = createPrebuiltRuleAssetsClient(savedObjectsClient);
+  const ruleObjectsClient = createPrebuiltRuleObjectsClient(rulesClient);
+
+  const prebuiltRulesMap = await fetchRuleVersionsTriad({
+    ruleAssetsClient,
+    ruleObjectsClient,
+  });
+
+  // Filter out prebuilt rules by `rule_id`
+  if (rulesIds) {
+    for (const ruleId of prebuiltRulesMap.keys()) {
+      if (!rulesIds.includes(ruleId)) {
+        prebuiltRulesMap.delete(ruleId);
+      }
+    }
+  }
```

Comment:
This looks overcomplicated:
```suggestion
  ruleIds?.forEach((ruleId) => prebuiltRulesMap.delete(ruleId));
```

Reply:
Actually, the one that you suggested won't work in this case. It will do opposite to what is expected :-) It will remove the prebuilt rules that are actually requested to be returned.

We could potentially replace this code with something like this if you think it will look better:

```
const existingIds = Array.from(prebuiltRulesMap.keys());
const filteredIds = existingIds.filter((id) => !rulesIds?.includes(id));
filteredIds.forEach((ruleId) => prebuiltRulesMap.delete(ruleId));
```

Reply:
Oh yes, you are right, I misunderstood the goal. 

Okay, what I don't quite like about this logic is that we have to iterate the `ruleIds` (_includes_) for every key in the `prebuiltRulesMap`, turning this into a quadratic complexity. 
We already have a Map available here with O(1) access by rule id, so we should be able to improve this to linear complexity. What about this:

```suggestion
  let filteredPrebuiltRulesMap;
  if (rulesIds) {
    filteredPrebuiltRulesMap = new Map();
    for (const ruleId of rulesIds) {
      const prebuiltRule = prebuiltRulesMap.get(ruleId);
      if (prebuiltRule) {
        filteredPrebuiltRulesMap.add(prebuiltRule);
      }
    }
  } else {
    filteredPrebuiltRulesMap = prebuiltRulesMap;
  }
```

So we only iterate `rulesIds` once.

Reply:
done